# ![alt text](https://drive.google.com/file/d/1Uw7VN5mObFG5NHaJFMEFMcNEHNeyWKPv/view?usp=drive_link)

# Low Input Epigenomics 2024 

### Course details: 
11 to 17 May 2024, Wellcome Connecting Science, Wellcome Genome Campus, Hinxton, 2024

Learn the latest approaches to study chromatin biology at the gene and genome-wide levels in rare cell populations

### Course Summary 
Our understanding of eukaryotic genomes has benefitted tremendously from whole genome sequencing projects. However, we are only beginning to systematically understand how the interplay of DNA, proteins and its biochemical modifications (i.e. chromatin) can influence global processes which control gene expression, mRNA splicing and DNA replication or repair in normal development and disease.

This hands-on practical course provides training in experimental and computational approaches that can be used to mechanically dissect important aspects of chromatin biology at the gene and genome-wide levels in rare cell populations.

Participants will receive intensive wet laboratory and computational training using various state-of-the-art methods that require fewer than 1000 cells input to bridge the gap between the standard bulk epigenomic approaches, often requiring inputs of millions of cells, and emerging single-cell techniques. This course will cover the wet lab and computational aspects of:

- Cleavage under targets and release using nucleases (CUT&RUN) to profile histone marks and transcription factor binding.
- Low-C to map the 3D structure of the genome and interactions between regulatory elements and gene promoters.
- Omni-ATAC to assess chromatin accessibility.
- Computational methods using various pipelines to analyse low-input genomic data and visualisation.

### Target audience

Senior PhD students, early career postdoctoral scientists, or junior faculty members who wish to interrogate epigenomic features on a low input scale.

### Programme
This 7 day course will cover the following techniques:

- (CUT&RUN) to profile histone marks and transcription factor binding.
- Low-C to map the 3D structure of the genome
- Omni-ATAC to assess chromatin accessibility.
- End-point analysis using qPCR and Tape station techniques
- Computational analysis of low input gene level and genome-wide data to determine the interplay of DNA, proteins and chromatin modifications in - biological processes. 
- These sessions will be supplemented with pre-course training on Linux and UNIX basics, informal tutorials, and guest lectures from the instructors and distinguished invited speakers.
- Participants will also have the opportunity to discuss their research interests and present aspects of their work relevant to the course.

### Learning Outcomes:

After attending the course, participants can expect to be able to:

- Perform CUT&RUN, Omni-ATAC, and Low-C techniques, including experimental design, endpoint analysis and validation (qPCR, TapeStation).
- Apply genome sequencing methods for low input genomic analysis
- Perform end-to-end analysis of low input genome sequence data from raw sequence to identifying loci of biological importance.
- Explore how low input genomic methods can be applied in development and disease research


### Training Team         

- [Noelia Diaz](https://www.researchgate.net/profile/Noelia-Diaz-6), Institute of Marine Sciences (ICM-CSIC), Spain
- [Kip Arlidge](https://www.linkedin.com/in/kiparlidge/), Canadore College, Canada
- [Kinjal Desai](https://www.linkedin.com/in/kinjal-desai-966371132/), The Hospital for Sick Children Research Institute and Cancer and Stem Cell Biology, Canada
- Noura Maziak, MRC-LMS London, UK
- Fabian Groll, The Institute of Cancer Research, UK
- [Benjamin Hernandez Rodriguez](https://ben-hr.github.io/), Max Planck Institute for Molecular Biomedicine, Germany


### Course Manual
**Module  - ATAC-seq**   
- [Online Manual - ATAC-seq](Modules/atac_seq_analysis/atac_analysis.md) 
        
- [PDF Manual - ATAC-seq](Modules/atac_seq_analysis/atac_analysis.pdf)   

        
                    
**Module Cut and Run**  
- [Online Manual - Cut and Run](Modules/cut_and_run_analysis_tutorial/CUT&RUN_analysis_tutorial.md)
        
- [PDF Manual - Cut and Run](Modules/cut_and_run_analysis_tutorial/CUT&RUN_analysis_tutorial.pdf)

    
                          
**Module Hi-C Matrix Generation**  
- [Online Manual - Hi-C Matrix Generation](Modules/hic_matrix_generation/hic_matrix_generation.md)

- [PDF Manual - Hi-C Matrix Generation](Modules/hic_matrix_generation/hic_matrix_generation.pdf)

         
                  
**Module HiC Feature Analysis**  
- [Online Manual - Hi-C Feature Analysis](Modules/hic_feature_analyses/hic_feature_analyses.md)

- [PDF Manual - Hi-C Feature Analysis](Modules/hic_feature_analyses/hic_feature_analyses.pdf)

         
                    

******
## License
<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">Attribution-NonCommercial-ShareAlike 4.0 International (CC BY-NC-SA 4.0)</a>.
